File: arena-jface/src/main/java/org/uqbar/lacar/ui/impl/jface/bindings/JFaceBindingBuilder.java
package org.uqbar.lacar.ui.impl.jface.bindings;

import org.eclipse.core.databinding.Binding;
import org.eclipse.core.databinding.DataBindingContext;
import org.eclipse.core.databinding.UpdateValueStrategy;
import org.eclipse.core.databinding.conversion.IConverter;
import org.eclipse.core.databinding.observable.value.IObservableValue;
import org.uqbar.arena.bindings.Transformer;
import org.uqbar.lacar.ui.impl.jface.JFaceWidgetBuilder;
import org.uqbar.lacar.ui.model.BindingBuilder;
import org.uqbar.ui.jface.base.BaseUpdateValueStrategy;

public class JFaceBindingBuilder implements BindingBuilder {
	private final DataBindingContext dbc;
	private IObservableValue model;
	private IObservableValue view;
	private UpdateValueStrategy viewToModel = new BaseUpdateValueStrategy();
	private UpdateValueStrategy modelToView = new BaseUpdateValueStrategy();

	public JFaceBindingBuilder(DataBindingContext dbc, IObservableValue view, IObservableValue model) {
		this.dbc = dbc;
		this.view = view;
		this.model = model;
	}

	public JFaceBindingBuilder(JFaceWidgetBuilder<?> widget, IObservableValue view, IObservableValue model) {
		this(widget.getDataBindingContext(), view, model);
	}

	/**
	 * Creates a {@link JFaceObservableFactory} whose model will be assigned later (using
	 * {@link #observeProperty(Object, String)}
	 * 
	 * @param widget
	 * @param view
	 */
	public JFaceBindingBuilder(JFaceWidgetBuilder<?> widget, IObservableValue view) {
		this(widget, view, null);
	}

	@Override
	public void observeProperty(Object model, String propertyName) {
		this.model = JFaceObservableFactory.observeProperty(model, propertyName);
	}

	protected void setConverter(UpdateValueStrategy viewToModel, IConverter converter) {
		viewToModel.setConverter(converter);
	}

	// ********************************************************
	// ** Configuration
	// ********************************************************

	@SuppressWarnings("unchecked")
	@Override
	public <M, V> BindingBuilder adaptWith(final Transformer<M, V> transformer) {
		this.setConverter(this.viewToModel, new IConverter() {
			@Override
			public Object getToType() {
				return transformer.getModelType();
			}

			@Override
			public Object getFromType() {
				return transformer.getViewType();
			}

			@Override
			public Object convert(Object value) {
				return transformer.viewToModel((V) value);
			}
		});

		this.setConverter(this.modelToView, new IConverter() {
			@Override
			public Object getToType() {
				return transformer.getViewType();
			}

			@Override
			public Object getFromType() {
				return transformer.getModelType();
			}

			@Override
			public Object convert(Object value) {
				return transformer.modelToView((M) value);
			}
		});

		return this;
	}

	// ********************************************************
	// ** Building
	// ********************************************************

	@Override
	public void build() {
		this.createBinding();
	}

	/**
	 * Creates the JFace data binding. Useful for adding behavior in subclasses that are aware of JFace.
	 * 
	 * @return The JFace binding object
	 */
	protected Binding createBinding() {
		return this.dbc.bindValue(view, model, viewToModel, modelToView);
	}
}

File: arena-jface/src/main/scala/org/uqbar/lacar/ui/impl/jface/bindings/JFaceBindingBuilder.scala
package org.uqbar.lacar.ui.impl.jface.bindings

import org.uqbar.lacar.ui.model.BindingBuilder
import org.eclipse.core.databinding.DataBindingContext
import org.eclipse.core.databinding.observable.value.IObservableValue
import org.uqbar.lacar.ui.impl.jface.builder.JFaceWidgetBuilder
import org.eclipse.core.databinding.conversion.IConverter
import org.eclipse.core.databinding.UpdateValueStrategy
import org.uqbar.arena.bindings.Transformer
import org.uqbar.ui.jface.base.BaseUpdateValueStrategy
import org.eclipse.core.runtime.IStatus
import org.uqbar.lacar.ui.impl.jface.builder.traits.JFaceContainer
import org.eclipse.core.databinding.observable.value.IObservableValue
import org.eclipse.core.runtime.IStatus
import org.uqbar.arena.jface.JFaceImplicits.closureToComputedValue
import org.uqbar.lacar.ui.model.BindingBuilder

/**
 *
 * @author npasserini
 * @author jfernandes
 */
class JFaceBindingBuilder(val dbc: DataBindingContext, val container: JFaceContainer, var view: IObservableValue, var model: IObservableValue) 
	extends BindingBuilder {
  
  private var viewToModel = new BaseUpdateValueStrategy()
  private var modelToView = new BaseUpdateValueStrategy()

  def this(widget: JFaceWidgetBuilder[_], view: IObservableValue, model: IObservableValue) {
    this(widget.getDataBindingContext, widget.container , view, model)
  }
  
  def this(widget: JFaceWidgetBuilder[_], view: IObservableValue) {
	  this(widget, view, null);
  }

  override def observeProperty(model: Object, propertyName: String) {
    this.model = JFaceObservableFactory observeProperty(model, propertyName)
  }

  override def observeErrors() = {
    model = () => container.status.getValue.asInstanceOf[IStatus].isOK
  }

  protected def setConverter(viewToModel: UpdateValueStrategy, converter: IConverter) {
    viewToModel setConverter converter
  }

  override def adaptWith[M, V](transformer: Transformer[M, V]) = {
    setConverter(viewToModel, new IConverter() {
      override def getToType() = transformer.getModelType
      override def getFromType() = transformer.getViewType
      override def convert(value: Object) = transformer.viewToModel(value.asInstanceOf[V]).asInstanceOf[Object]
    })

    setConverter(modelToView, new IConverter() {
      override def getToType() = transformer getViewType
      override def getFromType() = transformer getModelType
      override def convert(value: Object) = transformer.modelToView(value.asInstanceOf[M]).asInstanceOf[Object]
    })
    this
  }
  
  override def viewToModel[M,V](transformer : com.uqbar.commons.collections.Transformer[V,M]) : this.type = {
    setConverter(viewToModel, new IConverter() {
      override def convert(value:Object) = transformer.transform(value.asInstanceOf[V]).asInstanceOf[Object];
      //TODO: is this really needed ?
      override def getToType() = classOf[Object]
      override def getFromType() = classOf[Object]
    });
    this
  }
  
  override def modelToView[M,V](transformer : com.uqbar.commons.collections.Transformer[M,V]) : this.type = {
    setConverter(modelToView, new IConverter() {
      override def convert(value:Object) = transformer.transform(value.asInstanceOf[M]).asInstanceOf[Object];
      //TODO: is this really needed ?
      override def getToType() = classOf[Object]
      override def getFromType() = classOf[Object]
    });
    this
  }

  override def build() = createBinding()

  /**
   * Creates the JFace data binding. Useful for adding behavior in subclasses that are aware of JFace.
   *
   * @return The JFace binding object
   */
  protected def createBinding() = dbc bindValue (view, model, viewToModel, modelToView)

}